src/tree-manager/flatten-tree.js

```diff
@@ -139,39 +147,57 @@ function walkNodes({ nodes, list = new Map(), parent, depth = 0, simple, showPar
       node._id = node.id || `${i}`
     }
 
-    if (node.isDefaultValue) {
-      defaultValues.push(node._id)
-      node.checked = true
+    if (node.checked) {
```

Comment:
I think we should not need this and this is why:

Consider this test:

```
test('should select single first checked node and ignore any defaultValues', t => {
  const tree = [{ id: 'nodeA', isDefaultValue: true }, { id: 'nodeB', checked: true }, { id: 'nodeC', checked: true }]

  const manager = new TreeManager({ data: tree, radioSelect: true })

  t.false(manager.getNodeById('nodeA').checked)
  t.true(manager.getNodeById('nodeB').checked)
  t.false(manager.getNodeById('nodeC').checked)
})
```
First of all, 

```
  [{ id: 'nodeA', isDefaultValue: true }, { id: 'nodeB', checked: true }, { id: 'nodeC', checked: true }]
``` 

is bad input. If it is radio select, one should not be passing multiple `checked`, but since it can happen, we can assume it will.

Next, `defaultValue` trumps other nodes. This is because `defaultValues` are checked by default and thus, `nodeA` is the first checked node. And then, we're following first-person-wins strategy here, so the first defaultValue should get checked and nothing else. Of course this all made up stuff but we need to use our best judgement and decide one startegy or the other. If we're going to stick to the aforementioned strategy, then what we're saying is:

 - `defaultValue` should be checked regardless of `checked` prop (and this rule is currenlty enforced so it's kinda non-negotiable...)
 - in case of radio select, we're checking the first checked node we find and ignoring the rest (this is new and thus can be changed much more easily)

So based on that, the assertion `t.false(manager.getNodeById('nodeA').checked)` is incorrect. The correct assertion would be:

```
  t.true(manager.getNodeById('nodeA').checked)
  t.false(manager.getNodeById('nodeB').checked)
  t.false(manager.getNodeById('nodeC').checked)
```

But then, things become complicated if someone passes data in a different order:

```
  [{ id: 'nodeA', checked: true }, { id: 'nodeB', isDefaultValue: true }, { id: 'nodeC', checked: true }]

  // correct assertions based on rules
  t.false(manager.getNodeById('nodeA').checked)
  t.true(manager.getNodeById('nodeB').checked)
  t.false(manager.getNodeById('nodeC').checked)
```

In this case, tree manager would set `nodeA` checked as well as `nodeB`. It should remove `nodeA` when it encounters `nodeB` and I think this is why you need the `if (node.checked) {...` check to counter balance all these scenarios.

Remember that tree-manger is in the critical path and a) it needs to be as lean and fast as possible; b) having too many conditionals/branches increases cognitive complexity

I'm wondering if we can avoid this extra conditional if tweak things a little; e.g. 

 - create separate `walkNodes` for `simple`, `radio` and `hierarchial`
 - use iteration instrad of recursion **_if_** it helps simplfying things by keeping track of checked nodes, default values etc.

Maybe do this for radio select to begin with and then change for simple/hierachial afterwards.

And if all of this leads to a less simpler solution than what you have currently, then leave your implementation as is.

If you prefer, I can play around with these a bit and if you checked "allow contributors edits" while creating this PR, I can push to this for your review. Let me know either way.

Reply:
![bild](https://user-images.githubusercontent.com/17863113/54885802-6ba0ee00-4e80-11e9-973b-a7f5d7fb2266.png)
You should have green light for adding changes.

Reply:
In my opinion isDefaultValue should not trump a checked-state. Not after reading "the dropdown will automatically select the node(s) when there is no other selected node". I think it implies it's a fallback value, a.k.a no checked.

Most of the checks are quite simple and starts with mostly the less expensive condition so not sure how much of a difference it would make performance wise.

As I said it should be up for editing.

Reply:
> I think it implies it's a fallback value, a.k.a no checked.

Yes, but it still **_checks_** the input. Consider this - if someone provides `isDefaultValue: true, checked: false` - what would the checkbox do? Which prop do we honor (both on initial paint and after user actions of checking/unchecking other nodes)? 

This is what I meant by implied check - when we **init**, `isDefaultValue` implies it's checked and we ignore the checked prop.

Reply:
I've switched so isDefaultValue has precedence now. Have to anything regarding the iteration though.